community/kernel/src/test/java/org/neo4j/kernel/impl/storemigration/SchemaIndexMigratorTest.java

```diff
@@ -0,0 +1,247 @@
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.pagecache.PageCache;
+import org.neo4j.kernel.api.index.IndexAccessor;
+import org.neo4j.kernel.api.index.IndexConfiguration;
+import org.neo4j.kernel.api.index.IndexReader;
+import org.neo4j.kernel.api.index.SchemaIndexProvider;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.impl.api.index.sampling.IndexSamplingConfig;
+import org.neo4j.kernel.impl.store.AbstractStore;
+import org.neo4j.kernel.impl.store.NeoStore;
+import org.neo4j.kernel.impl.store.SchemaStore;
+import org.neo4j.kernel.impl.store.record.IndexRule;
+import org.neo4j.kernel.impl.store.record.SchemaRule;
+import org.neo4j.kernel.impl.storemigration.SchemaIndexMigrator.SchemaStoreProvider;
+import org.neo4j.kernel.impl.storemigration.StoreUpgrader.UnexpectedUpgradingStoreVersionException;
+import org.neo4j.kernel.impl.storemigration.legacystore.v19.Legacy19Store;
+import org.neo4j.kernel.impl.storemigration.legacystore.v20.Legacy20Store;
+import org.neo4j.kernel.impl.storemigration.legacystore.v21.Legacy21Store;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.neo4j.kernel.api.index.SchemaIndexProvider.getRootDirectory;
+
+public class SchemaIndexMigratorTest
+{
+    @Test
+    public void shouldNotMigrateFor1_9Version() throws Exception
+    {
+        // given
+        when( upgradableDatabase.hasCurrentVersion( fs, storeDir )).thenReturn( false );
+        when( upgradableDatabase.checkUpgradeable( storeDir ) ).thenReturn( Legacy19Store.LEGACY_VERSION );
+
+        // when
+        boolean needsMigration = migrator.needsMigration( storeDir );
+
+        // then
+        assertFalse( needsMigration );
+    }
+
+    @Test
+    public void shouldMigrateFor2_0Version() throws Exception
+    {
+        // given
+        when( upgradableDatabase.hasCurrentVersion( fs, storeDir )).thenReturn( false );
+        when( upgradableDatabase.checkUpgradeable( storeDir ) ).thenReturn( Legacy20Store.LEGACY_VERSION );
+
+        // when
+        boolean needsMigration = migrator.needsMigration( storeDir );
+
+        // then
+        assertTrue( needsMigration );
+    }
+
+    @Test
+    public void shouldMigrateFor2_1Version() throws Exception
+    {
+        // given
+        when( upgradableDatabase.hasCurrentVersion( fs, storeDir )).thenReturn( false );
+        when( upgradableDatabase.checkUpgradeable( storeDir ) ).thenReturn( Legacy21Store.LEGACY_VERSION );
+
+        // when
+        boolean needsMigration = migrator.needsMigration( storeDir );
+
+        // then
+        assertTrue( needsMigration );
+    }
+
+    @Test
+    public void shouldNotMigrateFor2_2Version() throws Exception
```

Comment:
@thobe I don't think it is possible to check that a given store has been created from a milestone release and then fail because of that. The only way to address that is to bump the store version, but that is exactly what I was trying to avoid.